### PR TITLE
FortiGate external portal module

### PR DESCRIPTION
**Pull request by fdurand**
# Description
Fortinet external portal support

# Delete branch after merge
YES

# NEWS file entries
## New Features
* Added support for external portal on FortiGate firewall

### CODE REVIEW

**Comment by extrafu** [comment]
@julsemaan Can you bring the logic behind the code from this PR into the new pfsso PR and close this one?

**Comment by julsemaan** [comment]
I would hapilly but this is not an SSO module, its a webauth module

**Comment by extrafu** [comment]
Oops you're right, sorry for the noise. @julsemaan please review it and merge it if all looks good code-wise.

File: lib/pf/Switch/Fortinet.pm [hunk 1]
```diff
@@ -0,0 +1,47 @@
+package pf::Switch::Fortinet;
+
+=head1 NAME
+
+pf::Switch::Fortinet - Object oriented module to access and configure enabled Netgear switches.
```
Comment (julsemaan) [inline lib/pf/Switch/Fortinet.pm]: Doesn't look like a Netjire switch module to me

File: lib/pf/Switch/Fortinet.pm [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+=cut
+
+use strict;
+use warnings;
+
+use Log::Log4perl;
+use Net::SNMP;
```
Comment (julsemaan) [inline lib/pf/Switch/Fortinet.pm]: Those two imports don't have any purpose from what I can see

File: lib/pf/Switch/Fortinet.pm [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+use pf::Switch::constants;
```
Comment (julsemaan) [inline lib/pf/Switch/Fortinet.pm]: useless import ?

File: lib/pf/Switch/Fortinet.pm [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+use base ('pf::Switch');
+
+=head1 AUTHOR
+
+Inverse inc. <info@inverse.ca>
+
+=head1 COPYRIGHT
+
+Copyright (C) 2005-2015 Inverse inc.
```
Comment (julsemaan) [inline lib/pf/Switch/Fortinet.pm]: 2017

File: lib/pf/Switch/Fortinet/FortiGate.pm [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+   return 'external';
+}
+
+
+=item parseExternalPortalRequest
+Parse external portal request using URI and it's parameters then return an hash reference with the appropriate parameters
+See L<pf::web::externalportal::handle>
+=cut
+
+sub parseExternalPortalRequest {
```
Comment (julsemaan) [inline lib/pf/Switch/Fortinet/FortiGate.pm]: Just to make sure, is this compatible with the rework @dwlfrth did ?

File: lib/pf/Switch/Fortinet/FortiGate.pm [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+        <script language="JavaScript" type="text/javascript">
+        window.setTimeout('document.weblogin_form.submit();', 1000);
+        </script>
+    ];
+
+    $logger->debug("Generated the following html form : ".$html_form);
+    return $html_form;
+}
+
+sub deauthenticateMacDefault {
```
Comment (julsemaan) [inline lib/pf/Switch/Fortinet/FortiGate.pm]: perldoc

File: lib/pf/Switch/Fortinet/FortiGate.pm [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+    get_logger->info("No doing deauthentication since this is a web form released switch.");
```
Comment (julsemaan) [inline lib/pf/Switch/Fortinet/FortiGate.pm]: Add this to the BUGS and LIMITATIONS in the perldoc at the top so we know

File: lib/pf/Switch/Fortinet/FortiGate.pm [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+=head1 NAME
+
+pf::Switch::Fortinet::FortiGate - Object oriented module to FortiGate using the external captive portal
+
+=head1 SYNOPSIS
+
+The pf::Switch::Fortinet::FortiGate  module implements an object oriented interface to interact with the FortiGate captive portal
+
+=head1 STATUS
+
```
Comment (julsemaan) [inline lib/pf/Switch/Fortinet/FortiGate.pm:12]: Please add the material and version on which it was tested

File: lib/pf/Switch/Fortinet/FortiGate.pm [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+sub description { 'FortiGate Firewall with web auth' }
+
+sub supportsExternalPortal { return $TRUE; }
+sub supportsWebFormRegistration { return $TRUE }
+sub supportsWirelessMacAuth { return $TRUE; }
+sub supportsWiredMacAuth { return $TRUE; }
+
+
+sub getIfIndexByNasPortId {
```
Comment (julsemaan) [inline lib/pf/Switch/Fortinet/FortiGate.pm]: perldoc

File: lib/pf/Switch/Fortinet/FortiGate.pm [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+}
+
+
+=head1 AUTHOR
+
+Inverse inc. <info@inverse.ca>
+
+=head1 COPYRIGHT
+
+Copyright (C) 2005-2015 Inverse inc.
```
Comment (julsemaan) [inline lib/pf/Switch/Fortinet/FortiGate.pm]: 2017

**Comment by extrafu** [comment]
Bump - can we wrap this up?

**Pull request merged**